### PR TITLE
added distinct branch for problem two

### PR DESCRIPTION
**Pull request by romeo-folie**

### CODE REVIEW

**Comment by kyonkopa** [comment]
See http://docs.php.net/manual/en/class.ds-map.php and use that as your data structure

File: activity.php [hunk 1]
```diff
@@ -12,15 +12,21 @@ public function __construct($data){
...
+            }
+        }
+
         // define contracts for basic fetch and mutate functionality
-        abstract public function getActivities();
+        abstract public function getActivities() : iterable;
         abstract public function getActivity($key);
         abstract public function addActivity($key, $value);
-        abstract public function deleteActivity($key);
+        abstract public function deleteActivity($key): iterable;
```
Comment (kyonkopa) [inline activity.php]: This should return boolean

File: activity.php [hunk 2]
```diff
@@ -12,15 +12,21 @@ public function __construct($data){
...
+        public function printData(){
+            foreach($this->data as $item => $value){
+                echo "{$item} - {$value} \n";
+            }
+        }
+
         // define contracts for basic fetch and mutate functionality
-        abstract public function getActivities();
+        abstract public function getActivities() : iterable;
         abstract public function getActivity($key);
```
Comment (kyonkopa) [inline activity.php]: Rename to `get`

File: activity.php [hunk 3]
```diff
@@ -12,15 +12,21 @@ public function __construct($data){
...
         abstract public function addActivity($key, $value);
```
Comment (kyonkopa) [inline activity.php]: Rename to `add`

File: activity.php [hunk 4]
```diff
@@ -12,15 +12,21 @@ public function __construct($data){
...
-        abstract public function deleteActivity($key);
```
Comment (kyonkopa) [inline activity.php:19]: Rename to `delete`

File: activity.php [hunk 5]
```diff
@@ -33,16 +39,17 @@ public function addActivity($key, $value) {
...
         }
 
-        public function deleteActivity($key){
+        public function deleteActivity($key): iterable {
             unset($this->data[$key]);
             return $this->data;
         }
     }
 
     $sampleData = array("eating" => "daily", "climbing" => "twice yearly", "coding" => "daily");
```
Comment (kyonkopa) [inline activity.php:55]: Initialise with an empty store

**Review by kyonkopa — CHANGES_REQUESTED** [review]

File: activity.php [hunk 6]
```diff
@@ -19,37 +22,31 @@ public function printData(){
...
-        abstract public function deleteActivity($key): iterable;
+        abstract public function get() : Ds\Sequence;
+        abstract public function add($key, $value);
+        abstract public function delete($key) : bool;
     }
 
     class ActivityImpl extends Activity {
-        public function getActivities(): iterable {
-            return $this->data;
+        public function get() : Ds\Sequence {
```
Comment (kyonkopa) [inline activity.php]: Accept `key` and return value in get. Then create an `all` method to return all entries

File: activity.php [hunk 7]
```diff
@@ -19,37 +22,31 @@ public function printData(){
...
+            return $this->data->pairs();
         }
 
-        public function getActivity($key) {
-            return $this->data[$key];
-        }
-
-        public function addActivity($key, $value) {
+        public function add($key, $value) {
```
Comment (kyonkopa) [inline activity.php:40]: Return value on `add` and use `Ds\Map::put` method

File: activity.php [hunk 8]
```diff
@@ -19,37 +22,31 @@ public function printData(){
...
             $this->data[$key] = $value;
-            return $this->data;
+            return $this->data[$key];
         }
 
-        public function deleteActivity($key): iterable {
+        public function delete($key) : bool {
```
Comment (kyonkopa) [inline activity.php:45]: Use `Ds\Map::remove` for this

File: activity.php [hunk 9]
```diff
@@ -19,37 +22,31 @@ public function printData(){
...
             unset($this->data[$key]);
-            return $this->data;
+            return $this->data->hasKey($key) ? false : true;
```
Comment (kyonkopa) [inline activity.php]: Return false if key was not found

**Review by kyonkopa — CHANGES_REQUESTED** [review]

File: activity.php [hunk 10]
```diff
@@ -1,48 +1,52 @@
...
     /* 
     Model an abstract Activity class that allows mutation and fetching of an Activity resource from a data store, 
     your class should define contracts for basic fetch and mutate functionality. 
     Then define an implementation of the abstraction. 
+    Use a Map as your data store
     */
 
     abstract class Activity {
-        public $data;
+        public $data = [];
```
Comment (kyonkopa) [inline activity.php]: This should be `protected`

**Review by kyonkopa — CHANGES_REQUESTED** [review]

File: activity.php (hunk 9 above)
Comment (kyonkopa) [reply activity.php]: What you want to do here is assume remove operation will pass if key exists however if key does not already exist return false

Comment (romeo-folie) [reply activity.php]: I added a try/catch to return false when the key isn't present in the map.

**Review by kyonkopa — APPROVED** [review]

**Pull request merged**